docs/HowToFAQ/FAQMapExtensionality.md

```diff
@@ -0,0 +1,114 @@
+
+```dafny
+MapKeepsCount(authSchema, k => Paths.SimpleCanon(tableName, k));
+assume |authSchema| == |map k <- authSchema.Keys | true :: Paths.SimpleCanon(tableName, k) := authSchema[k]|;
+assert |authSchema| == |map k <- authSchema.Keys | true :: Paths.SimpleCanon(tableName, k) := authSchema[k]|;
+```
+
+Which makes no sense.
+I even put the function into a variable, and this still fails
+```dafny
+assume |authSchema| == |map k <- authSchema.Keys :: fn(k) := authSchema[k]|;
+assert |authSchema| == |map k <- authSchema.Keys :: fn(k) := authSchema[k]|;
+```
+
+## Answer:
+
+The explanation is a little involved, and in the end gets into a weakness of Dafny. But I also have a workaround. Get some popcorn. Here goes:
+
+To prove that the `|map …|` expression in the specification has the same value as the `|map …|` expression in the code, 
+the verifier would either have to compute the cardinality of each map (which it can’t do, because m.Keys is symbolic and could stand for any size set) 
+or reason that the one map … is the very same map as the other map … (in which case it would follow that the cardinality of the two maps are also the same).
+The way to prove that two maps are equal is to show that they have the same keys and the same mappings. 
+The idea of proving two things equal by looking at the “elements” of each of the two things is called extensionality. 
+Dafny never tries to prove extensionality, but it’s happy to do it if you ask it to. 
+For example, if G is a function that you know nothing about and you ask to prove
+```dafny
+assert G(s + {x}) == G({x} + s + {x});
+```
+then Dafny will complain. You have to first establish that the arguments to these two invocations of `G` are the same, which you can do with an assert:
+```dafny
+assert s + {x} == {x} + s + {x};
+assert G(s + {x}) == G({x} + s + {x});
+```
+
+Here, Dafny will prove the first assertion (which it actually does by proving that the sets are elementwise equal) and will then use the first assertion to prove the second.
+Going back to your example, Dafny needs to know that the two `map`s are equal. To help it along, perhaps you could mention the two in an assertion, like
+
+`assert map … == map …;`
+
+But you can’t do that here, because the two map expressions are in different scopes and use different variables.
+To establish that the two maps are equal, we thus need to do something else. 
+If the two of them looked the same, then Dafny would know that the are the same. 
+But the form is slightly different, because you are (probably without thinking about it) instantiating a lambda expression. 
+To make them the same, you could rewrite the code to:
+```dafny
+  var F := k => Paths.SimpleCanon(tableName, k);
+  MapKeepsCount(authSchema, F);
+  assert |authSchema| == |map k <- authSchema.Keys | true :: F(k) := authSchema[k]|;
+```
+
+Now, this `map …` syntactically looks just like the one in the lemma postcondition, but with authSchema instead of m and with F instead of f. 
```

Comment:
```suggestion
Now, this `map …` syntactically looks just like the one in the lemma postcondition, but with `authSchema` instead of `m` and with `F` instead of `f`. 
```